atef/procedure.py

```diff
@@ -878,20 +924,335 @@ def from_origin(
+
+        validation_status = [status[0] for status in plan_status]
+        if not all(validation_status):
+            # raise an error, place info in result
+            fail_plan_names = [pl.name for pl, st in
+                               zip(self.prepared_plans, validation_status)
+                               if not st]
+            logger.debug(f'One or more plans ({fail_plan_names}) failed validation')
+            return Result(
+                severity=Severity.error,
+                reason=f'One or more plans ({fail_plan_names}) failed validation'
+            )
+
+        # send each plan to correct destination
+        plan_results = []
+        for pplan in self.prepared_plans:
+            logger.debug(f'running plan: {pplan.name}...')
+            res = await pplan.run()
+            logger.debug(f'run completed: {res}')
+            plan_results.append(res)
+
+        # run the checks
+        for prep_check in self.prepared_checks:
+            await prep_check.compare()
+
+        check_results = [check.result for check in self.prepared_checks]
+
+        if self.prepared_checks_failures:
+            return Result(
+                severity=Severity.error,
+                reason=('One or more success criteria failed to initialize: '
+                        f'{[check.name for check in self.prepared_checks_failures]}')
+            )
+
+        severity = _summarize_result_severity(GroupResultMode.all_,
+                                              check_results + plan_results)
+        reason = (f'{len(plan_results)} plans run, '
+                  f'{len(check_results)} checks passed')
+        return Result(severity=severity, reason=reason)
+
+    @classmethod
+    def from_origin(
+        cls,
+        origin: PlanStep,
+        parent: Optional[PreparedProcedureGroup] = None
+    ) -> PreparedPlanStep:
+
+        prep_step = cls(
+            origin=origin,
+            parent=parent,
+            name=origin.name,
+        )
+
+        for plan_step in origin.plans:
+            try:
+                prep_plan = PreparedPlan.from_origin(plan_step, parent=prep_step)
+                prep_step.prepared_plans.append(prep_plan)
+            except Exception:
+                prep_step.prepared_plan_failures.append(plan_step)
+
+        for check in origin.checks:
+            res = create_prepared_comparison(check, parent=prep_step)
+            if isinstance(res, Exception):
+                prep_step.prepared_checks_failures.append(res)
+            else:
+                prep_step.prepared_checks.append(res)
+
+        return prep_step
+
+
+def create_prepared_comparison(
+    check: Union[ComparisonToTarget, ComparisonToPlanData],
+    parent: Optional[AnyDataclass] = None
+) -> Union[PreparedComparison, PreparedComparisonException]:
+
+    output = ValueError('Cannot prepare the provided comparison, type not '
+                        f'supported: ({check})')
+
+    if isinstance(check, ComparisonToTarget):
```

Comment:
You previously mentioned the annoying isinstance if-blocks for these sorts of things, so let me mention something that came to mind: singledispatch
https://docs.python.org/3/library/functools.html#functools.singledispatch
It could be used to remove the `if`/`isinstance` block boilerplate here. I had used it (a total of once) in the pvaccess stuff in caproto and thought it was neat.

Reply:
I've never seen this single dispatch thing before, it's neat
shout-outs to the cursed match statement as well for isinstance if trees